src/integer_mod_q/mat_polynomial_ring_zq/arithmetic/add.rs

```diff
@@ -0,0 +1,278 @@
+    /// use qfall_math::integer_mod_q::ModulusPolynomialRingZq;
+    /// use qfall_math::integer::MatPolyOverZ;
+    /// use std::str::FromStr;
+    ///
+    /// let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+    /// let poly_mat1 = MatPolyOverZ::from_str("[[4  -1 0 1 1, 1  42],[0, 2  1 2]]").unwrap();
+    /// let poly_ring_mat1 = MatPolynomialRingZq::from((&poly_mat1, &modulus));
+    /// let poly_mat2 = MatPolyOverZ::from_str("[[3  3 0 1, 1  42],[0, 1  17]]").unwrap();
+    /// let poly_ring_mat2 = MatPolynomialRingZq::from((&poly_mat2, &modulus));
+    ///
+    /// let poly_ring_mat3: MatPolynomialRingZq = &poly_ring_mat1 + &poly_ring_mat2;
+    /// let poly_ring_mat4: MatPolynomialRingZq = poly_ring_mat1 + poly_ring_mat2;
+    /// let poly_ring_mat5: MatPolynomialRingZq = &poly_ring_mat3 + poly_ring_mat4;
+    /// let poly_ring_mat6: MatPolynomialRingZq = poly_ring_mat3 + &poly_ring_mat5;
+    /// ```
+    ///
+    /// # Panics ...
+    /// - ... if the moduli of both [`MatPolynomialRingZq`] mismatch.
+    fn add(self, other: Self) -> Self::Output {
+        self.add_safe(other).unwrap()
+    }
+}
+
+impl MatPolynomialRingZq {
+    /// Implements addition for two [`MatPolynomialRingZq`] values.
+    ///
+    /// Parameters:
+    /// - `other`: specifies the polynomial to add to `self`
+    ///
+    /// Returns the sum of both polynomials as a [`MatPolynomialRingZq`] or an error if the moduli
+    /// mismatch.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer_mod_q::MatPolynomialRingZq;
+    /// use qfall_math::integer_mod_q::ModulusPolynomialRingZq;
+    /// use qfall_math::integer::MatPolyOverZ;
+    /// use std::str::FromStr;
+    ///
+    /// let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+    /// let poly_mat1 = MatPolyOverZ::from_str("[[4  -1 0 1 1, 1  42],[0, 2  1 2]]").unwrap();
+    /// let poly_ring_mat1 = MatPolynomialRingZq::from((&poly_mat1, &modulus));
+    /// let poly_mat2 = MatPolyOverZ::from_str("[[3  3 0 1, 1  42],[0, 1  17]]").unwrap();
+    /// let poly_ring_mat2 = MatPolynomialRingZq::from((&poly_mat2, &modulus));
+    ///
+    /// let poly_ring_mat3: MatPolynomialRingZq = poly_ring_mat1.add_safe(&poly_ring_mat2).unwrap();
+    /// ```
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type [`MathError::MismatchingModulus`] if the moduli of
+    /// both [`MatPolynomialRingZq`] mismatch.
+    pub fn add_safe(&self, other: &Self) -> Result<MatPolynomialRingZq, MathError> {
+        if self.modulus != other.modulus {
+            return Err(MathError::MismatchingModulus(format!(
+                " Tried to add polynomial with modulus '{}' and polynomial with modulus '{}'.",
+                self.modulus, other.modulus
+            )));
+        }
+        let mut out = MatPolynomialRingZq::from((
+            &MatPolyOverZ::new(self.get_num_rows(), self.get_num_columns()).unwrap(),
```

Comment:
it would be nice, if there is an implementation, that takes this in as owned, because it would allow for a more efficient implementation. Please add a todo AND an issue/or task, that an owned from is implemented, because an entire copy would otherwise be done without need.

Should also be added for the PolyOverZ, but here it is an even greater loss of effciency. Fix it later, but add an issue, that we do not forget

Reply:
This is already implemented (see macros)